docs/pallas/gpu/pipelining.md

```diff
@@ -0,0 +1,325 @@
+a = jax.random.uniform(key1, shape=(m, k), dtype=jnp.float16)
+b = jax.random.uniform(key2, shape=(k, n), dtype=jnp.float16)
+
+result = matmul(a, b)
+
+np.testing.assert_allclose(result, a @ b)
+```
+
+<!-- #region id="lIYV7PN9J8Px" -->
+### Warp Specialization
+
+Warp specialization is a technique where we program each warp/warpgroup to perform a single task in order to give the GPU hardware the flexibility to schedule them at runtime. Recall that each streaming multiprocessor (SM) in a GPU contains warp schedulers that can swap execution between warps, so for example when one warp is blocked it can begin executing a different warp. In practice, this can be more performant than programming a single instruction stream and having the compiler statically schedule the operations to overlap the instructions them optimally.
+
+In particular, we are interested in warpgroup specialization on Hopper+ GPUs, where we have a separate warpgroup issuing TMAs (GMEM/SMEM copies) from the warpgroups performing arithmetic, since indexing calculations and issuing TMAs can take up a significant amount of time and potentially leave the TensorCore idle. The figure below depicts a standard, non-specialized kernel on the left where TMAs (async copies) and matrix multiplication are written to a single instruction stream, and a warp-specialized version on the right where communication and arithmetic are handled on separate warpgroups. A *consumed barrier* is used to synchronize between the specialized warpgroups that signals to the memory warpgroup when it is safe to begin the next TMA.
+
+
+<!-- #endregion -->
+
+<!-- #region id="n-y90IC7v7vL" -->
+
+![warp_specialization](../../_static/pallas/gpu/warp_specialization.svg)
+
+<!-- #endregion -->
+
+<!-- #region id="ZH0Pui5kFSdD" -->
+Warp specialization can be enabled in Pallas by using the `plgpu.emit_pipeline_warp_specialized` helper. This pipeline helper handles all of the logic in the memory warpgroup, and the user only needs to specify the work done in the compute warpgroups. It shares the a similar API as the standard `emit_pipeline`, and currently supports the following arguments:
+
+```python
+plgpu.emit_pipeline_warp_specialized(
+  body: Callable,
+  *
+  grid: tuple[int, ...],
+  in_specs: Sequence[pallas_core.BlockSpec] = (),
+  out_specs: Sequence[pallas_core.BlockSpec] = (),
+  max_concurrent_steps: int,
+  carry_coroutine: Callable
+  num_compute_wgs: int,
+  memory_registers: int
+  wg_axis: str,
+  memory_thread_idx: int | None = None,
+)
+```
+
+There are a few arguments specific to this pipeline emitter, which are:
+- `num_compute_wgs` specifies how many compute warpgroups to use. The pipeline emitter always uses a single memory warpgroup, so in `plgpu.kernel` you should specify `num_threads=num_compute_wgs+1`.
+- `memory_registers` controls how many registers to allocate to the memory warpgroup. The remaining registers are partitioned evenly among the compute warpgroups.
```

Comment:
Say that this can generally be left to be automatically inferred? Or at least give some rule of thumb for how many a user should specify?

Reply:
It's not automatically inferred though, we just default to 40. I just noted "should be adjusted up or down depending on whether register spills are encountered."